src/handlers/miner_hbbft_handler.erl

```diff
@@ -232,141 +245,149 @@ handle_command({txn, Txn}, State=#state{hbbft=HBBFT}) ->
+handle_message(<<BinMsgIn/binary>>, Index, #state{hbbft=HBBFT}=S0) ->
+    CurRound = hbbft:round(HBBFT),
+    case bin_to_msg(BinMsgIn) of
+        %% Mult sigs
+        {ok, {signatures, SigRound, _   }} when SigRound > CurRound ->
             defer;
-        {signatures, R, _Signatures} when R < Round ->
+        {ok, {signatures, SigRound, _   }} when SigRound < CurRound ->
             ignore;
-        {signatures, _R, Signatures} ->
-            Sigs = dedup_signatures(Signatures, State),
-            NewState = State#state{signatures = Sigs},
-            case enough_signatures(NewState) of
-                {ok, done, MySignatures} when Round > NewState#state.signed ->
+        {ok, {signatures, _, SigsIn}} ->
+            case state_sigs_input(SigsIn, S0) of
+                {{done, SigsOut}, S1} ->
                     ?mark(done_sigs),
-                    %% no point in doing this more than once
-                    ok = miner:signed_block(MySignatures, State#state.artifact),
-                    {NewState#state{signed = Round, sig_phase = done},
-                     [{multicast, t2b({signatures, Round, MySignatures})}]};
-                {ok, gossip, MySignatures} ->
-                    {NewState#state{sig_phase = gossip},
-                     [{multicast, t2b({signatures, Round, MySignatures})}]};
-                _ ->
-                    {NewState, []}
+                    MsgOut = msg_to_bin({signatures, CurRound, SigsOut}),
+                    {S1, [{multicast, MsgOut}]};
+                {{gossip, SigsOut}, S1} ->
+                    MsgOut = msg_to_bin({signatures, CurRound, SigsOut}),
+                    {S1, [{multicast, MsgOut}]};
+                {pending, S1} ->
+                    {S1, []}
             end;
-        {signature, R, Address, Signature} ->
-            case R == Round andalso lists:member(Address, State#state.members) andalso
-                 %% provisionally accept signatures if we don't have the means to verify them yet, they get filtered later
-                 (State#state.artifact == undefined orelse libp2p_crypto:verify(State#state.artifact, Signature, libp2p_crypto:bin_to_pubkey(Address))) of
-                true ->
-                    NewState = State#state{signatures=lists:keystore(Address, 1, State#state.signatures, {Address, Signature})},
-                    case enough_signatures(NewState) of
-                        {ok, done, Signatures} when Round > NewState#state.signed ->
+
+        %% Sing sig
+        {ok, {signature, SigRound, _, _}} when SigRound > CurRound ->
+            defer;
+        {ok, {signature, SigRound, _, _}} when SigRound < CurRound ->
+            ignore;
+        {ok, {signature, _, Addr, Sig}} ->
+            %% TODO Unify the 3 handlings of state_sigs_input output.
+            %% There 2 differences:
+            %% - call to ?mark(gossip_sigs) - 2 have it, 1 doesn't
+            %% - send signature on pending - 1 has it, 2 don't
```

Comment:
?mark(gossip_sigs) is fine to unify.

For sending the signature, we don't have it in all cases, so it cannot be unfied. You could pass in a signature | undefined for the different call sites though.

Reply:
Specifically we are sending our *own* signature, which we have just obtained over the artifact in that one case.

Reply:
> ?mark(gossip_sigs) is fine to unify.

It is called after inputting a single sig (i.e. `{signature, ...` msg), but not after inputting multiple sigs (i.e. `{signatures, ...` msg). Do we need to:
- a) preserve as is;
- b) call `?mark(gossip_sigs)` in both places;
- c) remove `?mark(gossip_sigs)` in both places.
?

Reply:
> For sending the signature, we don't have it in all cases, so it cannot be unfied. You could pass in a signature | undefined for the different call sites though.

Makes sense.

Reply:
> > ?mark(gossip_sigs) is fine to unify.
> 
> It is called after inputting a single sig (i.e. `{signature, ...` msg), but not after inputting multiple sigs (i.e. `{signatures, ...` msg). Do we need to:
> 
>     * a) preserve as is;
> 
>     * b) call `?mark(gossip_sigs)` in both places;
> 
>     * c) remove `?mark(gossip_sigs)` in both places.
>       ?

I think anywhere we transition to the gossip state.